builtin/v9/migration/verifreg.go

```diff
@@ -0,0 +1,143 @@
+package migration
+
+import (
+	"context"
+
+	verifreg8 "github.com/filecoin-project/go-state-types/builtin/v8/verifreg"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/builtin"
+	market8 "github.com/filecoin-project/go-state-types/builtin/v8/market"
+	adt8 "github.com/filecoin-project/go-state-types/builtin/v8/util/adt"
+	init9 "github.com/filecoin-project/go-state-types/builtin/v9/init"
+	market9 "github.com/filecoin-project/go-state-types/builtin/v9/market"
+	adt9 "github.com/filecoin-project/go-state-types/builtin/v9/util/adt"
+	verifreg9 "github.com/filecoin-project/go-state-types/builtin/v9/verifreg"
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+)
+
+func migrateVerifreg(ctx context.Context, adtStore adt8.Store, priorEpoch abi.ChainEpoch, initStateV9 init9.State, marketStateV8 market8.State, verifregStateV8 verifreg8.State, emptyMapCid cid.Cid) (cid.Cid, map[abi.DealID]verifreg9.AllocationId, error) {
+	pendingProposals, err := adt8.AsSet(adtStore, marketStateV8.PendingProposals, builtin.DefaultHamtBitwidth)
+	if err != nil {
+		return cid.Undef, nil, xerrors.Errorf("failed to load pending proposals: %w", err)
+	}
+
+	proposals, err := market8.AsDealProposalArray(adtStore, marketStateV8.Proposals)
+	if err != nil {
+		return cid.Undef, nil, xerrors.Errorf("failed to get proposals: %w", err)
+	}
+
+	pendingMap := make(map[abi.DealID]market8.DealProposal)
+	var proposal market8.DealProposal
+	if err = proposals.ForEach(&proposal, func(dealID int64) error {
+		pcid, err := proposal.Cid()
+		if err != nil {
+			return err
+		}
+
+		isPending, err := pendingProposals.Has(abi.CidKey(pcid))
+		if err != nil {
+			return xerrors.Errorf("failed to check pending: %w", err)
+		}
+
+		if isPending {
+			pendingMap[abi.DealID(dealID)] = proposal
+		}
+
+		return nil
+	}); err != nil {
+		return cid.Undef, nil, xerrors.Errorf("failed to iterate over proposals: %w", err)
+	}
+
+	nextAllocationID := verifreg9.AllocationId(1)
+	allocationsMapMap := make(map[address.Address]*adt9.Map)
+	dealsToAllocations := make(map[abi.DealID]verifreg9.AllocationId)
+	for dealID, proposal := range pendingMap {
+		clientIDAddress, ok, err := initStateV9.ResolveAddress(adtStore, proposal.Client)
+		if err != nil {
+			return cid.Undef, nil, xerrors.Errorf("failed to resolve client %s: %w", proposal.Client, err)
+		}
+
+		if !ok {
+			return cid.Undef, nil, xerrors.New("failed to find client in init actor map")
+		}
+
+		clientIDu64, err := address.IDFromAddress(clientIDAddress)
+		if err != nil {
+			return cid.Undef, nil, err
+		}
```

Comment:
Nit: there'd be good clarity gains from extracting pedestrian logic like this into functions so that the main migration code gets straight to the point.